packages/sanes-chrome-extension/src/logic/account/index.ts

```diff
@@ -1,24 +1,247 @@
+  // and ensure they have keys for all chains
+  public async ensureAllAccounts(): Promise<
+    ReadonlyArray<Account<AccountInfo>>
+  > {
+    const maybeAccounts = await this.loadAllAccounts();
+    return Promise.all(
+      maybeAccounts.map(maybe =>
+        this.ensureAccount(maybe.derivation, maybe.name)
+      )
+    );
+  }
+
+  // createChainAccount will create the new key based on the derivation.
+  // if it already exists it will throw an error (so check first)!
+  private async createChainAccount(
+    info: ChainDerivation,
+    derivation: number
+  ): Promise<AccountInfo> {
+    const { chainId, algorithm, derivePath, encoder } = info;
+    const path = derivePath(derivation);
+    const wallet = this.walletForAlgorithm(algorithm);
+    const identity = await this.profile.createIdentity(wallet, chainId, path);
+    return {
+      chainId,
+      algorithm,
+      derivation,
+      path,
+      identity,
+      address: encoder.identityToAddress(identity),
+    };
+  }
+
+  // loadChainAccount will check for account (there or missing) for one chain and derivation
+  private async loadChainAccount(
+    info: ChainDerivation,
+    derivation: number
+  ): Promise<MaybeAccountInfo> {
+    const { chainId, algorithm, derivePath, encoder } = info;
+    const path = derivePath(derivation);
+    const wallet = this.walletForAlgorithm(algorithm);
+    const identity = await this.checkPathExists(wallet, chainId, path);
+    if (identity === undefined) {
+      return {
+        chainId,
+        derivation,
+        missing: true,
+      };
+    }
+    return {
+      chainId,
+      algorithm,
+      derivation,
+      path,
+      identity,
+      address: encoder.identityToAddress(identity),
+    };
+  }
+
+  private walletForAlgorithm(algorithm: Algorithm): WalletId {
+    const [edWallet, secpWallet] = this.profile.wallets.value.map(x => x.id);
+    return algorithm === 'ed25519' ? edWallet : secpWallet;
+  }
+
+  /*
+  we really need a
+  public async checkIdentity(
+    walletId: WalletId,
+    chainId: ChainId,
+    options: Ed25519Keypair | ReadonlyArray<Slip10RawIndex> | number,
+  ): Promise<PublicIdentity>|undefined {
+    // returns the corresponding public identity if there already
+  }
+*/
+
+  // FIXME
+  // this is an ugly hack... this really needs to be added to the root UserProfile API
+  private async checkPathExists(
+    walletId: WalletId,
+    chainId: ChainId,
+    path: ReadonlyArray<Slip10RawIndex>
+  ): Promise<PublicIdentity | undefined> {
+    // Breaking encapsulation... bad Ethan
+    const wallet: ReadonlyWallet = (this
+      .profile as any).findWalletInPrimaryKeyring(walletId); // eslint-disable-line @typescript-eslint/no-explicit-any
+    const ident = await wallet.previewIdentity(chainId, path);
+    const allIdentities = wallet.getIdentities();
+    return allIdentities.find(x => publicIdentityEquals(x, ident));
```

Comment:
To replace this whole function, we would need a

```
  /**
   * Checks the existence of an identitiy in the wallet with the given ID in the primary keyring
   */
  public async identityExists(
    walletId: WalletId,
    chainId: ChainId,
    options: Ed25519Keypair | ReadonlyArray<Slip10RawIndex> | number,
  ): Promise<bool> {
    // ...
  }
```

i.e. the same interface as `UserProfile.createIdentity` but as an existence check, right?